OpenStack auth provider: persistent token caching

dlipovetsky opened the pull request:
**What this PR does / why we need it**:
Adds persistent caching of Keystone tokens to the OpenStack auth provider plugin.

**Which issue(s) this PR fixes**:
fixes #60541

TODO

- [x] Write unit tests

dlipovetsky committed:
Persistent token caching for the OpenStack auth provider

tallclair commented:
@kubernetes/sig-auth-pr-reviews

tallclair commented:
/ok-to-test

dlipovetsky commented:
My apologies--looks like I should have noted that the PR was not ready to test.

I did not have time recently, but am working on the tests this weekend.

dims commented:
looking forward to an update @dlipovetsky

dlipovetsky commented:
Finished up tests and rebased.

Thank you to the authors of the gcp auth provider, from which I borrowed ideas and code.

Also thanks to the authors of golang/oauth2. The `TestOpenstackTokenExpiresAt` code is slightly modified version of [a similar test in that package](https://github.com/golang/oauth2/blob/3c3a985cb79f52a3190fbc056984415ca6763d01/token_test.go#L35). If you think this is egregious, please let me know and I will re-write the test.

dlipovetsky commented:
(The "boilerplate check" failed for token_test.go because there was no newline between the boilerplate and the package declaration...)

dims commented:
@Fedosin can you please take a look?

@dlipovetsky Can you please squash the commits?

Fedosin commented:
@dims yep, I'm looking

Fedosin commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/openstack/openstack.go:
```diff
@@ -35,72 +36,121 @@ func init() {
+	ret := map[string]string{}
+	for k, v := range g.cache {
+		ret[k] = v
+	}
```
Is there a need to copy the cache into ret? Maybe we just can update and return g.cache directly?  Tbh I don't know why they did that for gcp https://github.com/kubernetes/client-go/blob/master/plugin/pkg/client/auth/gcp/gcp.go#L242-L244

dlipovetsky replied:
From what I understand, as soon as `update` returns, another goroutine can update `g.cache`. If we return `g.cache` directly, we may end up persisting the other goroutine's values, or garbage (e.g. our expires-at, but the other goroutine's token-id).

For more context, please see [the commit that introduced this](https://github.com/kubernetes/kubernetes/commit/5a6cd558c23657977ea4cb9103aa51d068c3fe57) to the gcp auth provider.

dlipovetsky commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/openstack/openstack.go:
```diff
@@ -35,72 +36,121 @@ func init() {
+	return token, nil
+}
+
+func (g *cachedGetter) cachedToken() *openstackToken {
```
Good point. Having `cachedToken()` return an error cleans up the method itself.

But I think the `Valid` method is correct as-is, since a nil token or one whose ID is an empty string is indeed invalid.

dlipovetsky commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/openstack/token.go:
```diff
@@ -0,0 +1,123 @@
+func tokenFromAuth(options gophercloud.AuthOptions) (*openstackToken, error) {
+	client, err := openstack.NewClient(options.IdentityEndpoint)
+	if err != nil {
+		return &openstackToken{}, fmt.Errorf("error creating openstack client: %s", err)
```
Thanks for catching this. Fixed.

dlipovetsky commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/openstack/token.go:
```diff
@@ -0,0 +1,123 @@
+		TenantID:         options.TenantID,
+		TenantName:       options.TenantName,
+		TokenID:          options.TokenID,
+	}
```
Thanks, will fix.

I initially thought `v2Opts` is declared because `tokens2.Create` would fail is passed v3-speciifc options like DomainID and DomainName. But now I see that [`ToTokenV2CreateMap` ](https://github.com/gophercloud/gophercloud/blob/d240260ea5048753e7c2f7afffd13ee6979ac179/openstack/identity/v2/tokens/requests.go#L56)takes care of this conversion internally.

dlipovetsky commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/openstack/openstack.go:
```diff
@@ -112,55 +162,30 @@ func (t *tokenRoundTripper) RoundTrip(req *http.Request) (*http.Response, error)
+	if res.StatusCode == 401 {
+		glog.V(4).Infof("The credentials that were supplied are invalid for the target cluster")
+		emptyCache := make(map[string]string)
+		t.persister.Persist(emptyCache)
```
Yes. Prior to this PR, the only value was `ttl`, which is no longer used. Would it be better to erase only the `token-id` and `expires-at` values?

dlipovetsky commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/openstack/token.go:
```diff
@@ -0,0 +1,123 @@
+type openstackToken struct {
+	ID        string
+	ExpiresAt time.Time
+}
```
That's not a bad thought. From what I understand, there are two ways to use the existing Token type. Let me know if I'm missing another one.
1. Replace `openstackToken` with `Token`. If I do this, I won't be able to define the `Valid` method, because `Token` is in a different package. So I think this is not an option.
2. Embed `Token` in `openstackToken`. I can do this, but constructing a literal openstackToken becomes cumbersome:
```
&openstackToken{
		ID:        id,
		ExpiresAt: expiresAtTime,
	}
```
vs
```
&openstackToken{&tokens3.Token{
		ID:        id,
		ExpiresAt: expiresAtTime,
	}}
```

dlipovetsky commented:
@Fedosin Thanks a lot for the review!
@dims I plan to squash the commits once I've addressed all the reviews. (If I squash now GitHub will lose track of the code to which the comments point.)

dlipovetsky commented:
@dims I addressed the review issues a few days ago and now I've squashed.

dims commented:
/approve

@Fedosin, if you please give a +1 after peeking at the squashed commit, then i can lgtm

thanks,
Dims

dims commented:
/assign @liggitt
